python/src/ir.cc

```diff
@@ -1603,6 +1606,10 @@ void init_triton_ir(py::module &&m) {
                                                llvm::StringRef(prefix));
              self.create<PrintOp>(prefixAttr, hex, values, isSigned);
            })
+      .def("create_proton_record",
```

Comment:
Let's take the python api changes out of this PR. It's not determined yet and probably not very important in the short term. Users, like @pawelszczerbuk, can modify the GPU IR to instrument record ops

Reply:
This is mostly just to have a template in place for the front end - we can always modify the interface later but how it all fit together with the Python frontend wasn't super intuitive at first so having this here I think is helpful for the moment even if we ultimately change the API in the future. It also gives us some more testing avenues. What do you think?

Reply:
The concern I have is that we don't have any real functionalities associated with these record ops anyway. Usually, we add a frontend op when there's at least some functionalities implemented. For example, the most recent `tl.gather` op. Though its initial commit doesn't address all problems and concerns, frontend users are able to use the op in their kernels but not just placeholders. So IMHO it makes more sense to add the op, and mark it as `experimental`, after the lowering has been implemented.

Reply:
What about putting things in a `dev` module with some warnings about them being under development and not intentioned for use outside the development team:

https://github.com/triton-lang/triton/pull/5506/files#diff-054c0ddf64263bf99d77a237a61deeb79c0cd3e4289a03eba948ce833f1cdce0R17

https://github.com/triton-lang/triton/pull/5506/files#diff-bc216e7a68b5a4545f1ff69bfbf5d0fa64adb6b1acaca6ea0f366f47e93942ebR25

Then that gives us an easy method to develop/test in a way that turning them "on" is just a matter of moving them from the `dev` module to the `experimental` one?

Reply:
It's much better now!

Let's just move it under `proton.language` without using the dev class.

Reply:
Also I think `create_proton_record` can be moved to `triton_proton.cc`

Reply:
I think this would cause a lot of code disruption. TritonOpBuilder is a unique ptr wrapper around the MLIR OpBuilder. So I think no matter what we'd have to find way to pass that object to triton_proton.cc either through moving the definition of TritonOpBuilder out of ir.cc into a header file or possibly through a helper function defined in ir.cc and called from triton_proton.cc which seems like an equal amount of complexity.

Reply:
I get it now. Let's move it to the last line of all ops and leave a comment:

```
// proton ops
.def("create_proton_record",
...
```

It's a problem for all third party backends that they cannot define custom ops. We will have to think about a better solution.

Reply:
Updated. 

> It's a problem for all third party backends that they cannot define custom ops. We will have to think about a better solution.

Right, there has to be an upper level Triton Op that the backends can specialize but I could not find a way to have a wholly custom third party backend Op